README
NAME
    IPC::ConcurrencyLimit - Lock-based limits on cooperative
    multi-processing

SYNOPSIS
      use IPC::ConcurrencyLimit;
  
      sub run {
        my $limit = IPC::ConcurrencyLimit->new(
          type      => 'Flock', # that's also the default
          max_procs => 10,
          path      => '/var/run/myapp', # an option to the locking strategy
        );
    
        my $id = $limit->get_lock;
        if (not $id) {
          warn "Got none of the worker locks. Exiting.";
          exit(0);
        }
        else {
          # Got one of the worker locks (ie. number $id)
          do_work();
        }
    
        # lock released with $limit going out of scope here
      }
  
      run();
      exit();

DESCRIPTION
    This module implements a mechanism to limit the number of concurrent
    processes in a cooperative multiprocessing environment. This is an
    alternative to, for example, running several daemons.

    Roughly speaking, a typical setup would be the following:

    * Cron starts a new process every minute.

    * The process attempts to get a lock as shown in synopsis.

    * If it obtains a lock, it starts working and exits when the work is
      done.

    * If not, "max_procs" processes are already working, so it exits.

    This has several distinct advantages over daemons.

    * Processes do not run as long. Small memory leaks are less likely to
      become a problem.

    * Rolling out new code is trivial. No need to do any daemon restarting
      and worrying about interrupting a unit of work.

    * No complicated master/slave setups and process/thread pooling.

    The implementation uses some form of locking to limit concurrency:
    There's simply a limited number of locks to go around. The detailed
    locking implementation is chosen using the "type" parameter to the
    constructor. The base distributions ships with one locking strategy
    only: IPC::ConcurrencyLimit::Lock::Flock for a file-locking based
    concurrency limit.

    Among the other potential strategies that are not part of this
    distribution are NFS-based locking using File::SharedNFSLock or using
    MySQL's "GET_LOCK". Both of these schemes would allow limiting
    concurrency across multiple hosts without a special-purpose daemon.

METHODS
  new
    Creates a new concurrency limit. Creating the object does not lock
    anything. This requires a followup call to the "get_lock()" method!

    After calling "get_lock()", the lock will be held by the
    "IPC::ConcurrencyLimit" object and released when either "release_lock()"
    is called or the "IPC::ConcurrencyLimit" object is freed.

    Takes named parameters, one of which is the "type" parameter, which
    specifies the type of lock to use and defaults to "Flock".

    The "max_procs" option indicates the maximum number of locks that can be
    held at the same time and thus usually the maximum no. of running
    processes. It defaults to 1.

    All concurrency limits that refer to the same resource/limit must use
    the same setting for "max_procs". If not, the behaviour is undefined.

    All other named parameters will be passed as options to the lock
    implementation. See IPC::ConcurrencyLimit::Lock::Flock or other
    implementations.

  get_lock
    Creates the actual lock and if successful, returns its id (starting from
    1, not 0).

    Returns undef if locking was unsuccessful. Multiple calls do not stack.
    They will simply return the same lock id as long as a lock is held.

  release_lock
    Releases the lock. Returns 1 if a lock has been released or undef if
    there had been no lock to be released.

  is_locked
    Returns whether we have a lock.

  lock_id
    Returns the id of the lock or undef if there is none.

AUTHOR
    Steffen Mueller, "smueller@cpan.org"

    Yves Orton

ACKNOWLEDGMENT
    This module was originally developed for booking.com. With approval from
    booking.com, this module was generalized and put on CPAN, for which the
    authors would like to express their gratitude.

COPYRIGHT AND LICENSE
     (C) 2011 Steffen Mueller. All rights reserved.
 
     This code is available under the same license as Perl version
     5.8.1 or higher.
 
     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



Changes
Revision history for Perl extension IPC::ConcurrencyLimit

0.16 Thu Aug 25 16:29:09 CEST 2016
  - default process_name_change in WithLatestStandby to off.
  - use constant memory instead of scaling to max_procs
    during execution in IPC::ConcurrencyLimit::Flock.

0.15  Wed Aug 24 14:16 2016
  - add new WithLatestStandby backend (YVES)
  - bugfix related to retry count (YVES)

0.14  Tue Sep 22 10:01 2015
  - No-op release to fix META.json+META.yml+Changes
    (https://rt.cpan.org/Public/Bug/Display.html?id=107204)

0.13  Fri Sep 18 15:04 2015
  - added warning in the POD about lock variable scope
  - Flock:
    - start at the *highest* id, and count *down*
      to find the lock we want to use.
    - add file_prefix option
    - add file_ext option

0.12  Thu Jan 29 15:20 2015
  - Use `retries` callback earlier

0.11  Tue Aug 20 21:42 2013
  - Fix incorrect documentation for retries parameter

0.10  Fri Oct 12 17:00 2012
  - The C<heartbeat> method of IPC::ConcurrencyLimit and
    IPC::ConcurrencyLimit::Lock object can now be used to check
    whether a lock was lost since it was first acquired.

0.09  Fri Oct  5 12:08 2012
  - IPC::ConcurrencyLimit::WithStandby now has dynamically
    configurable retries via a callback mechanism.

0.08  Tue Oct  2 09:47 2012
  - Fix IPC::ConcurrencyLimit::WithStandby synopsis.

0.07  Wed Jun 27 07:55 2012
  - Support modify_process_name in IPC::ConcurrencyLimit::WithStandby.

0.06  Mon Jun 25 09:25 2012
  - Add explicit loading of IO::File

0.05  Mon Feb  6 07:52 2012
  - Skip the forking tests for WithStandby on uncommon OS.
    Specifically, CPAN testers showed failures on Solaris that I have
    no hope of debugging. It's more likely to be a bug in the tests
    than a bug in the module. Contributors welcome.
    For now, disable forking tests on non-linux, non-darwin, non-win32.

0.04  Fri Feb  3 19:55 2012
  - Implement IPC::ConcurrencyLimit::WithStandby

0.03  Thu May 26 13:08 2011
  - Add missing flush to make sure PID is written to file.
    (Should not be a critical fix)

0.02  Tue Mar 19 12:10 2011
  - Expose shared-locking mode of operation.

0.01  Wed Mar 16 14:29 2011
  - original version


lib/IPC/ConcurrencyLimit.pm
package IPC::ConcurrencyLimit;
use 5.008001;
use strict;
use warnings;

our $VERSION = '0.16';

use Carp qw(croak);

sub new {
  my $class = shift;
  my %params = @_;
  my $type = delete $params{type};
  $type = 'Flock' if not defined $type;

  my $lock_class = $class . "::Lock::$type";
  if (not eval "require $lock_class; 1;") {
    my $err = $@ || 'Zombie error';
    croak("Invalid lock type '$type'. Could not load lock class '$lock_class': $err");
  }

  my $self = bless {
    opt => {
      max_procs => 1,
      %params,
    },
    lock_class => $lock_class,
    lock_obj => undef,
  } => $class;

  return $self;
}

sub get_lock {
  my $self = shift;
  return $self->{lock_obj}->id() if $self->{lock_obj};
  
  my $class = $self->{lock_class};
  $self->{lock_obj} = $class->new($self->{opt});

  return $self->{lock_obj} ? $self->{lock_obj}->id() : undef;
}

sub is_locked {
  my $self = shift;
  return $self->{lock_obj} ? 1 : 0;
}

sub release_lock {
  my $self = shift;
  return undef if not $self->{lock_obj};
  $self->{lock_obj} = undef;
  return 1;
}

sub lock_id {
  my $self = shift;
  return undef if not $self->{lock_obj};
  return $self->{lock_obj}->id;
}

sub heartbeat {
  my $self = shift;
  my $lock = $self->{lock_obj};
  return if not $lock;
  if (not $lock->heartbeat) {
    $self->release_lock;
    return();
  }
  return 1;
}

1;

__END__


=head1 NAME

IPC::ConcurrencyLimit - Lock-based limits on cooperative multi-processing

=head1 SYNOPSIS

  use IPC::ConcurrencyLimit;
  
  sub run {
    my $limit = IPC::ConcurrencyLimit->new(
      type      => 'Flock', # that's also the default
      max_procs => 10,
      path      => '/var/run/myapp', # an option to the locking strategy
    );
    
    # NOTE: when $limit goes out of scope, the lock is released
    my $id = $limit->get_lock;
    if (not $id) {
      warn "Got none of the worker locks. Exiting.";
      exit(0);
    }
    else {
      # Got one of the worker locks (ie. number $id)
      do_work();
    }
    
    # lock released with $limit going out of scope here
  }
  
  run();
  exit();

=head1 DESCRIPTION

This module implements a mechanism to limit the number of concurrent
processes in a cooperative multiprocessing environment. This is an alternative
to, for example, running several daemons.

Roughly speaking,
a typical setup would be the following:

=over 2

=item *

Cron starts a new process every minute.

=item *

The process attempts to get a lock as shown in synopsis.

=item *

If it obtains a lock, it starts working and exits when the
work is done.

=item *

If not, C<max_procs> processes are already working, so it exits.

=back

This has several distinct advantages over daemons.

=over 2

=item *

Processes do not run as long. Small memory leaks are less likely to
become a problem.

=item *

Rolling out new code is trivial. No need to do any daemon restarting
and worrying about interrupting a unit of work.

=item *

No complicated master/slave setups and process/thread pooling.

=back

The implementation uses some form of locking to limit concurrency:
There's simply a limited number of locks to go around. The detailed
locking implementation is chosen using the C<type> parameter to the
constructor. The base distributions ships with one locking strategy only:
L<IPC::ConcurrencyLimit::Lock::Flock> for a file-locking based
concurrency limit.

Among the other potential strategies that are not part of
this distribution are NFS-based locking using L<File::SharedNFSLock>
or using MySQL's C<GET_LOCK>. Both of these schemes would
allow limiting concurrency across multiple hosts without a
special-purpose daemon.

=head1 METHODS

=head2 new

Creates a new concurrency limit. Creating the object
does B<not> lock anything. This requires a followup
call to the C<get_lock()> method!

After calling C<get_lock()>, the lock will be held
by the C<IPC::ConcurrencyLimit> object and released
when either C<release_lock()> is called or the
C<IPC::ConcurrencyLimit> object is freed.

Takes named parameters, one of which is the C<type>
parameter, which specifies the type of lock to use
and defaults to C<Flock>.

The C<max_procs> option indicates the maximum
number of locks that can be held at the same time
and thus usually the maximum no. of running processes.
It defaults to 1.

All concurrency limits that refer to the same resource/limit
B<must> use the same setting for C<max_procs>. If not, the
behaviour is undefined.

All other named parameters will be passed as options
to the lock implementation. See L<IPC::ConcurrencyLimit::Lock::Flock>
or other implementations.

=head2 get_lock

Creates the actual lock and if successful, returns its id
(starting from 1, not 0).

Returns undef if locking was unsuccessful.
Multiple calls do not stack. They will simply return
the same lock id as long as a lock is held.

B<WARNING!> make sure the variable holding the lock remains in scope at all
times, otherwise the lock will be released and locking will become apparently
ineffective. This is the most common reason for having several concurrent
processes running when only one is expected to be alive.

=head2 release_lock

Releases the lock. Returns 1 if a lock has been
released or undef if there had been no lock to be
released.

=head2 is_locked

Returns whether we have a lock.

=head2 lock_id

Returns the id of the lock or undef if there is none.

=head2 heartbeat

Check whether the lock is still valid. If so,
returns true. Otherwise, it releases (destroys) the lock
and returns false.

=head1 AUTHOR

Steffen Mueller, C<smueller@cpan.org>

Yves Orton

David Morel

Matt Koscica, C<mattk@cpan.org>

Ivan Kruglov

=head1 ACKNOWLEDGMENT

This module was originally developed for Booking.com.
With approval from Booking.com, this module was generalized
and put on CPAN, for which the authors would like to express
their gratitude.

=head1 COPYRIGHT AND LICENSE

 (C) 2011-2015 Steffen Mueller. All rights reserved.
 
 This code is available under the same license as Perl version
 5.8.1 or higher.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut



lib/IPC/ConcurrencyLimit/Lock.pm
package IPC::ConcurrencyLimit::Lock;
use 5.008001;
use strict;
use warnings;

our $VERSION = '0.16';
sub id { $_[0]->{id} }

sub heartbeat { 1 }

1;

__END__


=head1 NAME

IPC::ConcurrencyLimit::Lock - Simple base class for lock implementations

=head1 SYNOPSIS

  use IPC::ConcurrencyLimit;

=head1 DESCRIPTION

Very simple base class for locks defining a common interface.

If you are just looking to use L<IPC::ConcurrencyLimit>,
you don't need this.

=head1 METHODS

=head2 new

Constructor that acquires a new lock and
returns undef on failure to acquire a lock.

Needs to be implemented in the subclass.

First argument is a hash reference containing
options including at least C<max_procs>
which indicates the maximum number of locks
at the same time.

I<Note to implementors:> Copy the hash if you need to store it.

=head2 DESTROY

The destructor (none implemented in the base class)
needs to release the lock.

=head2 id

Returns the id of the lock (starting at 1, not 0).

=head2 heartbeat

When called, must return whether the lock is still valid.
By default, this just returns true unless overridden in
subclasses.

=head1 AUTHOR

Steffen Mueller, C<smueller@cpan.org>

Yves Orton

=head1 ACKNOWLEDGMENT

This module was originally developed for booking.com.
With approval from booking.com, this module was generalized
and put on CPAN, for which the authors would like to express
their gratitude.

=head1 COPYRIGHT AND LICENSE

 (C) 2011, 2012 Steffen Mueller. All rights reserved.
 
 This code is available under the same license as Perl version
 5.8.1 or higher.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut



lib/IPC/ConcurrencyLimit/Lock/Flock.pm
package IPC::ConcurrencyLimit::Lock::Flock;
use 5.008001;
use strict;
use warnings;
use Carp qw(croak);
use File::Path qw();
use File::Spec;
use Fcntl qw(:DEFAULT :flock);
use IO::File ();

our $VERSION = '0.16';
use IPC::ConcurrencyLimit::Lock;
our @ISA = qw(IPC::ConcurrencyLimit::Lock);

sub new {
  my $class = shift;
  my $opt = shift;

  my $max_procs = $opt->{max_procs}
    or croak("Need a 'max_procs' parameter");
  my $path = $opt->{path}
    or croak("Need a 'path' parameter");
  my $lock_mode = lc($opt->{lock_mode} || 'exclusive');
  if ($lock_mode !~ /^(?:exclusive|shared)$/) {
    croak("Invalid lock mode '$lock_mode'");
  }
  my $file_prefix= defined $opt->{file_prefix} ? $opt->{file_prefix} : "";
  $file_prefix .= "." if length $file_prefix;
  my $file_ext= defined $opt->{file_ext} ? $opt->{file_ext} : "lock";
  $file_ext=~s/^\.?/./;

  my $self = bless {
    max_procs => $max_procs,
    path      => $path,
    lock_fh   => undef,
    lock_file => undef,
    id        => undef,
    file_prefix => $file_prefix,
    file_ext    => $file_ext,
    lock_mode => $lock_mode,
  } => $class;

  $self->_get_lock() or return undef;

  return $self;
}

sub _get_lock {
  my $self = shift;

  File::Path::mkpath($self->{path});
  my $lock_mode_flag = $self->{lock_mode} eq 'shared' ? LOCK_SH : LOCK_EX;

  # We try in reverse order, so that if a processor is started with
  # a higher number of allowed locks there is less chance that it starves
  # a processor with a lower number of allowed locks.
  my $worker_id = $self->{max_procs};
  while ($worker_id > 0) {
    my $lock_file = File::Spec->catfile($self->{path}, join("", $self->{file_prefix}, $worker_id, $self->{file_ext}));

    sysopen(my $fh, $lock_file, O_RDWR|O_CREAT)
      or die "can't open '$lock_file': $!";

    if (flock($fh, $lock_mode_flag|LOCK_NB)) {
      $self->{lock_fh} = $fh;
      seek($fh, 0, 0);
      truncate($fh, 0);
      print $fh $$;
      $fh->flush;
      $self->{id} = $worker_id;
      $self->{lock_file} = $lock_file;
      last;
    }

    close $fh;
    $worker_id--;
  }

  return undef if not $self->{id};
  return 1;
}

sub lock_file { $_[0]->{lock_file} }
sub path { $_[0]->{path} }

sub DESTROY {
  my $self = shift;
  # should be superfluous
  close($self->{lock_fh}) if $self->{lock_fh};
}

1;

__END__


=head1 NAME

IPC::ConcurrencyLimit::Lock::Flock - flock() based locking

=head1 SYNOPSIS

  use IPC::ConcurrencyLimit;

=head1 DESCRIPTION

This locking strategy implements C<flock()> based concurrency control.
Requires that your system has a sane C<flock()> implementation as well
as a non-blocking C<flock()> mode.

Inherits from L<IPC::LimitConcurrency::Lock>.

Take care not to attempt to use this on an NFS share or any other file
system that does not implement atomic C<flock()>!

=head1 METHODS

=head2 new

Given a hash ref with options, attempts to obtain a lock in
the pool. On success, returns the lock object, otherwise undef.

Required options:

=over 2

=item C<path>

The directory that will hold the lock files.
Created if it does not exist.
It is suggested not to use a directory that may hold other data unless you also
use the file_prefix option and ensure that your locks are unique.

=item C<file_prefix>

An optional prefix which will be prepended to lock file names. If specified then
files will be created that look like this:

    $path/$file_prefix.1.lock

=item C<file_ext>

By default lock files have a ".lock" suffix/extension. Set this option to specify
a different extension.

=item C<max_procs>

The maximum no. of locks (and thus usually processes)
to allow at one time.

=back

Other options:

=over 2

=item C<lock_mode>

Defaults to C<exclusive> locks.

In particular circumstance, you might want to set this to C<shared>.
This subverts the way the normal concurrency limit works, but allows
entirely different use cases.

=back

=head2 lock_file

Returns the full path and name of the lock file.

=head2 path

Returns the directory in which the lock files resides.

=head1 AUTHOR

Steffen Mueller, C<smueller@cpan.org>

Yves Orton

=head1 ACKNOWLEDGMENT

This module was originally developed for booking.com.
With approval from booking.com, this module was generalized
and put on CPAN, for which the authors would like to express
their gratitude.

=head1 COPYRIGHT AND LICENSE

 (C) 2011, 2012 Steffen Mueller. All rights reserved.
 
 This code is available under the same license as Perl version
 5.8.1 or higher.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut



lib/IPC/ConcurrencyLimit/WithStandby.pm
package IPC::ConcurrencyLimit::WithStandby;
use 5.008001;
use strict;
use warnings;

our $VERSION = '0.15';
use Carp qw(croak);
use Time::HiRes qw(sleep);
use IPC::ConcurrencyLimit;

sub new {
  my $class = shift;
  my %params = @_;
  my $type = delete $params{type};
  $type = 'Flock' if not defined $type;

  my $standby_type = delete($params{standby_type}) || $type;

  foreach my $t ($type, $standby_type) {
    my $lock_class = "IPC::ConcurrencyLimit::Lock::$t";
    if (not eval "require $lock_class; 1;") {
      my $err = $@ || 'Zombie error';
      croak("Invalid lock type '$t'. Could not load lock class '$lock_class': $err");
    }
  }

  my %standby;
  foreach my $key (grep /^standby_/, keys %params) {
    my $munged = $key;
    $munged =~ s/^standby_//;
    $standby{$munged} = delete $params{$key};
  }
  $standby{$_} = $params{$_} for grep !exists($standby{$_}), keys %params;

  my $main_lock = IPC::ConcurrencyLimit->new(%params, type => $type);
  my $standby_lock = IPC::ConcurrencyLimit->new(%standby, type => $standby_type);

  my $self = bless({
    main_lock => $main_lock,
    standby_lock => $standby_lock,
    retries => defined($params{retries}) ? $params{retries} : 10,
    interval => defined($params{interval}) ? $params{interval} : 1,
    process_name_change => $params{process_name_change},
  } => $class);

  return $self;
}

sub get_lock {
  my $self = shift;
  my $main_lock = $self->{main_lock};

  # Convert retries to a sub if it's not one already
  if ( ref $self->{retries} ne "CODE" ) {
      my $max_retries = $self->{retries};
      $self->{retries} = sub { return $_[0] <= $max_retries };
  }

  my $id = $main_lock->get_lock;
  return $id if defined $id;

  my $st_lock = $self->{standby_lock};
  my $st_id = $st_lock->get_lock;
  return undef if not defined $st_id;

  # got standby lock, go into wait-retry loop
  my $old_proc_name;
  if ($self->{process_name_change}) {
    $old_proc_name = $0;
    $0 = "$0 - standby";
  }
  my $interval = $self->{interval};
  eval {
    my $tries = 0;
    while (1) {
      $id = $main_lock->get_lock;
      if (defined $id) {
        $st_lock->release_lock;
        last;
      }

      last unless $self->{retries}->(++$tries);
      sleep($interval) if $interval;
    }
    1;
  }
  or do {
    my $err = $@ || 'Zombie error';
    $0 = $old_proc_name if defined $old_proc_name;
    $st_lock->release_lock;
    die $err;
  };

  $0 = $old_proc_name if defined $old_proc_name;
  return $id;
}

sub is_locked {
  my $self = shift;
  return $self->{main_lock}->is_locked(@_);
}

sub release_lock {
  my $self = shift;
  return $self->{main_lock}->release_lock(@_);
}

sub lock_id {
  my $self = shift;
  return $self->{main_lock}->lock_id(@_);
}

sub heartbeat {
  my $self = shift;
  return $self->{main_lock}->heartbeat;
}

1;

__END__


=head1 NAME

IPC::ConcurrencyLimit::WithStandby - IPC::ConcurrencyLimit with an additional standby lock

=head1 SYNOPSIS

  use IPC::ConcurrencyLimit::WithStandby;
  
  sub run {
    my $limit = IPC::ConcurrencyLimit::WithStandby->new(
      type              => 'Flock', # that's also the default
      max_procs         => 10,
      path              => '/var/run/myapp',
      standby_path      => '/var/run/myapp/standby',
      standby_max_procs => 3,
    );
    
    # NOTE: when $limit goes out of scope, the lock is released
    my $id = $limit->get_lock;
    if (not $id) {
      warn "Got none of the worker locks. Exiting.";
      exit(0);
    }
    else {
      # Got one of the worker locks (ie. number $id)
      do_work();
    }
    
    # lock released with $limit going out of scope here
  }
  
  run();
  exit();

=head1 DESCRIPTION

This module provides the same interface as the regular L<IPC::ConcurrencyLimit>
module. It differs in what happens if C<get_lock> fails to get a slot for the
main limit:

If it fails to get a (or the) lock on the main limit, it will repeatedly attempt
to get the main lock until a slot for the main limit is attained or the number
of retries is exhausted. Most importantly, this supports limiting the number of
instances that continuously attempt to get the main lock (typically, this would
be limited to 1). This is implemented with a wait-retry-loop and two separate
C<IPC::ConcurrencyLimit> objects.

The options for the main limit are passed
in to the constructor as usual. The standby limit are inherited from the main
one, but all parameters prefixed with C<standby_> will override the respective
inherited parameters. For example, C<standby_type =E<gt> "MySQL"> will
enforce the use of the MySQL lock for the standby lock.

In addition to the regular C<IPC::ConcurrencyLimit> options, the constructor
accepts C<retries> as the number of retries a standby instance should do
to get the main lock. There will always be only one attempt to become
a standby process. Additionally, C<interval> can indicate a number of seconds
to wait between retries (also supports fractional seconds down to what
C<Time::HiRes::sleep> supports).

C<retries> can also be passed a code reference that will be called on every
retry, with the current attempt number as its first argument (starting at 1).
Returning false from this routine will break the loop and give up any further
attempts to acquire the lock. One example of a configuration which would
continue to attempt to acquire a lock forever would be as follows:

  my $limit = IPC::ConcurrencyLimit::WithStandby->new(
    retries           => sub { $_[0] },
    interval          => 0.01,
    maxproc           => 1,
    standby_max_procs => 1,
    ...
  );

The form above would be used to have a single process running, with a standby
process ready to take over 1/100th of a second after the active process exits,
at the expense of attempting to acquire a C<flock()> 100 times per second.

Finally, as a way to tell blocked worker processes apart from standby
processes, the module supports the C<process_name_change> option. If set to
true, then the module will modify the process name of standby processes via
modification of <$0>. It appends the string " - standby" to $0 and resets it
to the old value after timing out or getting a worker lock. This is only
supported on newer Perls and might not work on all operating systems.
On my testing Linux, a process that showed as C<perl foo.pl> in the process
table before using this feature was shown as C<foo.pl - standby> while
in standby mode and as C<foo.pl> after getting a main worker lock.
Note the curiously stripped C<perl> prefix.

=head1 WARNING!

Make sure the variable holding the lock remains in scope at all times,
otherwise the lock will be released, resulting in process churn and potential
double work. This is the most common reason for having several concurrent
processes running when only one is expected to be alive.

=head1 AUTHOR

Steffen Mueller, C<smueller@cpan.org>

David Morel, C<david.morel@amakuru.net>

=head1 ACKNOWLEDGMENT

This module was originally developed for booking.com.
With approval from booking.com, this module was generalized
and put on CPAN, for which the authors would like to express
their gratitude.

=head1 COPYRIGHT AND LICENSE

 (C) 2012 Steffen Mueller. All rights reserved.

 This code is available under the same license as Perl version
 5.8.1 or higher.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut

